File: app/lib/teams.py
"""Team operations that are worth testing on their own."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Optional

from app.db import db_lock, get_db
from app.lib.ids import generate_access_token, generate_short_code
from app.lib.shuffle import shuffle_into_groups


@dataclass
class ShuffleResult:
    created: int
    placed: int
    deleted: int
    protected_teams: int
    skipped: int = 0


def copy_teams(from_assignment_id: int, to_assignment_id: int) -> int:
    """Clone one assignment's teams onto another. Returns how many were
    created.

    Fresh links and codes are minted on purpose -- the copy is a new set of
    teams, and a code that unlocked last week's delivery must not unlock
    this week's.

    A student already grouped for the target assignment is skipped rather
    than moved, so running this twice, or after some students have
    organised themselves, never undoes what is already there. A source team
    whose members are all already placed produces no empty team.
    """
    if from_assignment_id == to_assignment_id:
        return 0

    with db_lock() as conn:
        source_teams = conn.execute(
            "SELECT * FROM teams WHERE assignment_id = ?", (from_assignment_id,)
        ).fetchall()

        already_grouped = {
            row["student_id"]
            for row in conn.execute(
                "SELECT student_id FROM team_members WHERE assignment_id = ?",
                (to_assignment_id,),
            ).fetchall()
        }

        created = 0
        for team in source_teams:
            members = [
                row["student_id"]
                for row in conn.execute(
                    "SELECT student_id FROM team_members WHERE team_id = ?",
                    (team["id"],),
                ).fetchall()
                if row["student_id"] not in already_grouped
            ]
            if not members:
                continue

            cur = conn.execute(
                "INSERT INTO teams (assignment_id, name, access_token, short_code) "
                "VALUES (?, ?, ?, ?)",
                (to_assignment_id, team["name"], generate_access_token(), generate_short_code()),
            )
            new_team_id = cur.lastrowid

            conn.executemany(
                "INSERT INTO team_members (team_id, assignment_id, student_id) "
                "VALUES (?, ?, ?)",
                [(new_team_id, to_assignment_id, sid) for sid in members],
            )

            already_grouped.update(members)
            created += 1

        conn.commit()
        return created


def count_protected_teams(assignment_id: int) -> int:
    """How many of this assignment's teams a "re-shuffle everyone" would
    refuse to touch, because they already have a submission or a deadline
    extension. Used to name the real number in the confirm dialog rather
    than a hypothetical.
    """
    conn = get_db()
    ids = set()
    for row in conn.execute(
        "SELECT DISTINCT team_id FROM submissions WHERE assignment_id = ?",
        (assignment_id,),
    ).fetchall():
        ids.add(row["team_id"])
    for row in conn.execute(
        "SELECT DISTINCT team_id FROM deadline_extensions WHERE assignment_id = ?",
        (assignment_id,),
    ).fetchall():
        ids.add(row["team_id"])
    return len(ids)


def shuffle_teams(
    assignment_id: int,
    mode: str = "fill",
    preferred: int = 4,
    seed: Optional[int] = None,
) -> ShuffleResult:
    """Auto-group a roster into teams of a chosen size for one assignment.

    `mode="fill"` (the default) only touches students who have no team for
    this assignment yet, so running it after some students have organised
    themselves -- or running it twice -- never undoes what is already
    there.

    `mode="reshuffle"` re-cuts everyone, deleting the assignment's existing
    teams first -- except any team that already has a submission or a
    deadline extension attached. Both of those cascade off `teams.id`, so
    deleting such a team would silently take real deliveries and their
    files with it; this function refuses to do that regardless of what the
    caller asked for, since it may be reached directly rather than only
    through the confirm dialog that names the risk.
    """
    with db_lock() as conn:
        existing_teams = conn.execute(
            "SELECT * FROM teams WHERE assignment_id = ?", (assignment_id,)
        ).fetchall()

        protected_ids = set()
        for row in conn.execute(
            "SELECT DISTINCT team_id FROM submissions WHERE assignment_id = ?",
            (assignment_id,),
        ).fetchall():
            protected_ids.add(row["team_id"])
        for row in conn.execute(
            "SELECT DISTINCT team_id FROM deadline_extensions WHERE assignment_id = ?",
            (assignment_id,),
        ).fetchall():
            protected_ids.add(row["team_id"])

        protected_teams = [t for t in existing_teams if t["id"] in protected_ids]

        deleted = 0
        if mode == "reshuffle":
            removable = [t for t in existing_teams if t["id"] not in protected_ids]
            for t in removable:
                conn.execute("DELETE FROM teams WHERE id = ?", (t["id"],))
            deleted = len(removable)

            protected_student_ids = set()
            if protected_teams:
                placeholders = ",".join("?" for _ in protected_teams)
                rows = conn.execute(
                    f"SELECT student_id FROM team_members WHERE assignment_id = ? "
                    f"AND team_id IN ({placeholders})",
                    (assignment_id, *[t["id"] for t in protected_teams]),
                ).fetchall()
                protected_student_ids = {r["student_id"] for r in rows}

            pool = [
                r["id"]
                for r in conn.execute(
                    "SELECT id FROM students WHERE active = 1"
                ).fetchall()
                if r["id"] not in protected_student_ids
            ]
        else:
            already_grouped = {
                r["student_id"]
                for r in conn.execute(
                    "SELECT student_id FROM team_members WHERE assignment_id = ?",
                    (assignment_id,),
                ).fetchall()
            }
            pool = [
                r["id"]
                for r in conn.execute(
                    "SELECT id FROM students WHERE active = 1"
                ).fetchall()
                if r["id"] not in already_grouped
            ]

        groups = shuffle_into_groups(pool, preferred=preferred, seed=seed)

        next_number = 1
        for t in existing_teams:
            m = re.match(r"^Group (\d+)$", t["name"])
            if m:
                next_number = max(next_number, int(m.group(1)) + 1)

        created = 0
        placed = 0
        for group in groups:
            cur = conn.execute(
                "INSERT INTO teams (assignment_id, name, access_token, short_code) "
                "VALUES (?, ?, ?, ?)",
                (
                    assignment_id,
                    f"Group {next_number}",
                    generate_access_token(),
                    generate_short_code(),
                ),
            )
            next_number += 1
            team_id = cur.lastrowid

            conn.executemany(
                "INSERT INTO team_members (team_id, assignment_id, student_id) "
                "VALUES (?, ?, ?)",
                [(team_id, assignment_id, sid) for sid in group],
            )

            created += 1
            placed += len(group)

        conn.commit()

        return ShuffleResult(
            created=created,
            placed=placed,
            deleted=deleted,
            protected_teams=len(protected_teams),
        )

File: src/lib/teams.ts
import { and, eq, inArray } from "drizzle-orm";

import { db } from "@/db";
import {
  deadlineExtensions,
  students,
  submissions,
  teamMembers,
  teams,
} from "@/db/schema";
import { generateAccessToken, generateShortCode } from "./ids";
import { shuffleIntoGroups } from "./shuffle";

export interface ShuffleResult {
  created: number;
  placed: number;
  deleted: number;
  protectedTeams: number;
  skipped: number;
}

/**
 * Team operations that are worth testing on their own.
 *
 * Deliberately not a `"use server"` module: those may only export async
 * functions, and keeping this plain means the copy can be exercised directly
 * against a database instead of only through an HTTP action.
 */

/**
 * Clone one assignment's teams onto another. Returns how many were created.
 *
 * Fresh links and codes are minted on purpose — the copy is a new set of teams,
 * and a code that unlocked last week's delivery must not unlock this week's.
 *
 * A student already grouped for the target assignment is skipped rather than
 * moved, so running this twice, or after some students have organised
 * themselves, never undoes what is already there. A source team whose members
 * are all already placed produces no empty team.
 */
export function copyTeams(
  fromAssignmentId: number,
  toAssignmentId: number,
): number {
  if (fromAssignmentId === toAssignmentId) return 0;

  const source = db.query.teams
    .findMany({
      where: eq(teams.assignmentId, fromAssignmentId),
      with: { members: true },
    })
    .sync();

  const alreadyGrouped = new Set(
    db
      .select({ studentId: teamMembers.studentId })
      .from(teamMembers)
      .where(eq(teamMembers.assignmentId, toAssignmentId))
      .all()
      .map((r) => r.studentId),
  );

  let created = 0;
  db.transaction((tx) => {
    for (const team of source) {
      const members = team.members
        .map((m) => m.studentId)
        .filter((id) => !alreadyGrouped.has(id));
      if (members.length === 0) continue;

      const copy = tx
        .insert(teams)
        .values({
          assignmentId: toAssignmentId,
          name: team.name,
          accessToken: generateAccessToken(),
          shortCode: generateShortCode(),
        })
        .returning()
        .get();

      tx.insert(teamMembers)
        .values(
          members.map((studentId) => ({
            teamId: copy.id,
            assignmentId: toAssignmentId,
            studentId,
          })),
        )
        .run();

      // Guard against a source that somehow lists the same student twice.
      for (const id of members) alreadyGrouped.add(id);
      created++;
    }
  });

  return created;
}

/**
 * How many of this assignment's teams a "re-shuffle everyone" would refuse to
 * touch, because they already have a submission or a deadline extension. Used
 * to name the real number in the confirm dialog rather than a hypothetical.
 */
export function countProtectedTeams(assignmentId: number): number {
  const teamIds = new Set([
    ...db
      .select({ teamId: submissions.teamId })
      .from(submissions)
      .where(eq(submissions.assignmentId, assignmentId))
      .all()
      .map((r) => r.teamId),
    ...db
      .select({ teamId: deadlineExtensions.teamId })
      .from(deadlineExtensions)
      .where(eq(deadlineExtensions.assignmentId, assignmentId))
      .all()
      .map((r) => r.teamId),
  ]);
  return teamIds.size;
}

/**
 * Auto-group a roster into teams of 3-4 for one assignment.
 *
 * `mode: "fill"` (the default) only touches students who have no team for
 * this assignment yet, so running it after some students have organised
 * themselves — or running it twice — never undoes what is already there.
 *
 * `mode: "reshuffle"` re-cuts everyone, deleting the assignment's existing
 * teams first — except any team that already has a submission or a deadline
 * extension attached. Both of those cascade off `teams.id`, so deleting such
 * a team would silently take real deliveries and their files with it; this
 * function refuses to do that regardless of what the caller asked for, since
 * it may be reached directly rather than only through the confirm dialog that
 * names the risk.
 */
export function shuffleTeams(
  assignmentId: number,
  opts?: { mode?: "fill" | "reshuffle"; preferred?: number; seed?: number },
): ShuffleResult {
  const mode = opts?.mode ?? "fill";

  return db.transaction((tx) => {
    const existingTeams = tx
      .select()
      .from(teams)
      .where(eq(teams.assignmentId, assignmentId))
      .all();

    const protectedTeamIds = new Set([
      ...tx
        .select({ teamId: submissions.teamId })
        .from(submissions)
        .where(eq(submissions.assignmentId, assignmentId))
        .all()
        .map((r) => r.teamId),
      ...tx
        .select({ teamId: deadlineExtensions.teamId })
        .from(deadlineExtensions)
        .where(eq(deadlineExtensions.assignmentId, assignmentId))
        .all()
        .map((r) => r.teamId),
    ]);

    const protectedTeams = existingTeams.filter((t) =>
      protectedTeamIds.has(t.id),
    );

    let deleted = 0;
    let pool: number[];

    if (mode === "reshuffle") {
      const removable = existingTeams.filter(
        (t) => !protectedTeamIds.has(t.id),
      );
      for (const t of removable) {
        tx.delete(teams).where(eq(teams.id, t.id)).run();
      }
      deleted = removable.length;

      const protectedStudentIds = new Set(
        protectedTeams.length === 0
          ? []
          : tx
              .select({ studentId: teamMembers.studentId })
              .from(teamMembers)
              .where(
                and(
                  eq(teamMembers.assignmentId, assignmentId),
                  inArray(
                    teamMembers.teamId,
                    protectedTeams.map((t) => t.id),
                  ),
                ),
              )
              .all()
              .map((m) => m.studentId),
      );

      pool = tx
        .select({ id: students.id })
        .from(students)
        .where(eq(students.active, true))
        .all()
        .map((s) => s.id)
        .filter((id) => !protectedStudentIds.has(id));
    } else {
      const alreadyGrouped = new Set(
        tx
          .select({ studentId: teamMembers.studentId })
          .from(teamMembers)
          .where(eq(teamMembers.assignmentId, assignmentId))
          .all()
          .map((r) => r.studentId),
      );

      pool = tx
        .select({ id: students.id })
        .from(students)
        .where(eq(students.active, true))
        .all()
        .map((s) => s.id)
        .filter((id) => !alreadyGrouped.has(id));
    }

    const groups = shuffleIntoGroups(pool, {
      preferred: opts?.preferred ?? 4,
      seed: opts?.seed,
    });

    let nextNumber =
      1 +
      existingTeams.reduce((max, t) => {
        const match = /^Group (\d+)$/.exec(t.name);
        return match ? Math.max(max, Number(match[1])) : max;
      }, 0);

    let created = 0;
    let placed = 0;
    for (const group of groups) {
      const team = tx
        .insert(teams)
        .values({
          assignmentId,
          name: `Group ${nextNumber++}`,
          accessToken: generateAccessToken(),
          shortCode: generateShortCode(),
        })
        .returning()
        .get();

      tx.insert(teamMembers)
        .values(
          group.map((studentId) => ({
            teamId: team.id,
            assignmentId,
            studentId,
          })),
        )
        .run();

      created++;
      placed += group.length;
    }

    return {
      created,
      placed,
      deleted,
      protectedTeams: protectedTeams.length,
      skipped: 0,
    };
  });
}
